pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/SchemaConformingTransformerV2.java

```diff
@@ -0,0 +1,715 @@
+      return extraFieldsContainer;
+    }
+
+    String keyJsonPath = String.join(".", jsonPath);
+    Set<String> fieldPathsToDrop = _transformerConfig.getFieldPathsToDrop();
+    if (null != fieldPathsToDrop && fieldPathsToDrop.contains(keyJsonPath)) {
+      return extraFieldsContainer;
+    }
+
+    SchemaTreeNode currentNode = parentNode == null ? null : parentNode.getChild(key);
+    String unindexableFieldSuffix = _transformerConfig.getUnindexableFieldSuffix();
+    isIndexable = isIndexable && (null == unindexableFieldSuffix || !key.endsWith(unindexableFieldSuffix));
+    if (!(value instanceof Map)) {
+      // leaf node
+      if (!isIndexable) {
+        extraFieldsContainer.addUnindexableEntry(key, value);
+      } else {
+        if (null != currentNode && currentNode.isColumn()) {
+          // In schema
+          outputRecord.putValue(currentNode.getColumnName(), currentNode.getValue(value));
+          if (_transformerConfig.getFieldsToDoubleIngest().contains(keyJsonPath)) {
+            extraFieldsContainer.addIndexableEntry(key, value);
+          }
+          mergedTextIndexMap.put(keyJsonPath, value);
+        } else {
+          // Out of schema
+          if (storeIndexableExtras) {
+            extraFieldsContainer.addIndexableEntry(key, value);
+            mergedTextIndexMap.put(keyJsonPath, value);
+          }
+        }
+      }
+      return extraFieldsContainer;
+    }
+    // Traverse the subtree
+    Map<String, Object> valueAsMap = (Map<String, Object>) value;
+    for (Map.Entry<String, Object> entry : valueAsMap.entrySet()) {
+      jsonPath.addLast(entry.getKey());
+      ExtraFieldsContainer childContainer =
+          processField(currentNode, jsonPath, entry.getValue(), isIndexable, outputRecord, mergedTextIndexMap);
+      extraFieldsContainer.addChild(key, childContainer);
+      jsonPath.removeLast();
+    }
+    return extraFieldsContainer;
+  }
+
+  /**
+   * Generate an index token based on the provided key-value pair.
+   * The index token follows this format: "val:key".
+   * @param kv                            used to generate text index tokens
+   * @param indexTokens                   a list to store the generated index tokens
+   * @param mergedTextIndexTokenMaxLength which we enforce via truncation during token generation
+   */
+  public void generateTextIndexToken(Map.Entry<String, Object> kv, List<String> indexTokens,
+      Integer mergedTextIndexTokenMaxLength) {
+    String key = kv.getKey();
+    String val;
+    // To avoid redundant leading and tailing '"', only convert to JSON string if the value is a list or an array
+    if (kv.getValue() instanceof Collection || kv.getValue() instanceof Object[]) {
+      try {
+        val = JsonUtils.objectToString(kv.getValue());
+      } catch (JsonProcessingException e) {
+        val = kv.getValue().toString();
+      }
+    } else {
+      val = kv.getValue().toString();
+    }
+
+    if (key.length() + 1 > MAXIMUM_LUCENE_TOKEN_SIZE) {
```

Comment:
To be precise, "LUCENE TOKEN SIZE" is number of bytes and there could be UTF8 variable length characters. So `key.length() + 1`, which refers to # of characters, could overflow and cause indexing issues. If we want to make this check perfect, it's probably a good idea to encode the string first.

Note that his is a minor issue and to fix it properly, it requires changes in many places within this transformer.

Reply:
Good point. Mark it as a TODO. Not resolve in this patch because the upstream Kafka event usually has size limit and usually could not reach that high.